pennylane_cirq/qsim_device.py

```diff
@@ -0,0 +1,110 @@
+            or strings (``['ancilla', 'q1', 'q2']``). Default 1 if not specified.
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. Shots need
+            to be >= 1. In analytic mode, shots indicates the number of entries
+            that are returned by ``device.sample``.
+        analytic (bool): indicates whether expectation values and variances should
+            be calculated analytically
+        qubits (List[cirq.Qubit]): A list of Cirq qubits that are used
+            as wires. The wire number corresponds to the index in the list.
+            By default, an array of ``cirq.LineQubit`` instances is created.
+    """
+    name = "QSim device for PennyLane"
+    short_name = "cirq.qsim"
+
+    def __init__(self, wires, shots=1000, analytic=True, qubits=None):
+        super().__init__(wires, shots, analytic, qubits)
+        self.circuit = None
+        self._simulator = qsimcirq.QSimSimulator()
+
+    def reset(self):
+        # pylint: disable=missing-function-docstring
+        super().reset()
+        self.circuit = qsimcirq.QSimCircuit(cirq_circuit=self.circuit)
+
+
+class QSimhDevice(SimulatorDevice):
+    r"""QSimh device for PennyLane.
+
+    Args:
+        wires (int, Iterable[Number, str]]): Number of subsystems represented by the device,
+            or iterable that contains unique labels for the subsystems as numbers (i.e., ``[-1, 0, 2]``)
+            or strings (``['ancilla', 'q1', 'q2']``). Default 1 if not specified.
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. Shots need
+            to be >= 1. In analytic mode, shots indicates the number of entries
+            that are returned by ``device.sample``.
+        analytic (bool): indicates whether expectation values and variances should
+            be calculated analytically
+        qubits (List[cirq.Qubit]): A list of Cirq qubits that are used
+            as wires. The wire number corresponds to the index in the list.
+            By default, an array of ``cirq.LineQubit`` instances is created.
+        qsimh_options (dict): A dictionary with options for the QSimh simulator. See the `QSim
+            usage documentation <https://github.com/quantumlib/qsim/blob/master/docs/usage.md>`__
+            for further details.
+    """
+    name = "QSimh device for PennyLane"
+    short_name = "cirq.qsimh"
+
+    def __init__(self, wires, shots=1000, analytic=True, qubits=None, qsimh_options=None):
+        super().__init__(wires, shots, analytic, qubits)
+        if qsimh_options is None:
+            qsimh_options = {"k": [0], "w": 0, "p": 0, "r": 1}
+        self.circuit = None
+        self.qsimh_options = qsimh_options
+        self._simulator = qsimcirq.QSimhSimulator(qsimh_options)
+
+    def apply(self, operations, **kwargs):
+        CirqDevice.apply(self, operations, **kwargs)
```

Comment:
Just for my understanding: there is something in the `SimulatorDevice` version of `apply()` that we don't want to use, preventing us from calling `super().apply()` here?

Would it make sense for this device to inherit directly from `CirqDevice`? (maybe not, perhaps `SimulatorDevice` is useful in other places).

Reply:
Yes, that's exactly the case. `SimulatorDevice.apply` uses the method `_simulator.simulate` which qsimh doesn't have (qsim has it, and thus simply uses `SimulatorDevice.apply`). 

> Would it make sense for this device to inherit directly from CirqDevice? (maybe not, perhaps SimulatorDevice is useful in other places).

Not really, since quite a lot of the methods in `SimulatorDevice` are needed (e.g. `Cirq.apply` calls several abstract methods that only exist in `SimulatorDevice`). The alternative would be to make qsimh inherit from `CirqDevice` and rewrite many of the functions from `SimulatorDevice`, but I'm not sure if I think that actually makes things better. It might seem more structured, but would mean maintaining two version of many of the functions. I was thinking about this while writing it, and decided on the current solution. :slightly_smiling_face:

Reply:
Perfect, thanks for the explanation!